File: EightQueen.cpp
#include <iostream>
#include <cmath>

using namespace std;

const int n = 8;
int a[8];
int answers = 0;
bool exist[8];
bool back = 0;

bool ok(int r, int c) {
	for (int i = 0; i < r; ++i) {
		if (a[i] == c || (abs(c - a[i]) == abs(r - i))) return false;
	}
	return true;
}

int main() {
	for (int i = 0; i < n; ++i) {
		a[i] = -1;
		exist[i] = 0;
	}
	int r = 0, c = 0;
	while (r < n) {
		while (c < n) {
			if (ok(r, c)) {
				a[r] = c;
				c = 0;
				break;
			}
			else ++c;
		}
		if (a[r] == -1) {
			if (r == 0) break;
			else {
				--r;
				c = a[r] + 1;
				a[r] = -1;
				continue;
			}
		}
		if (r == n - 1) {
			++answers;
			cout << "\nThe NO. " << answers << " answer\n";
			cout << "row\tcolumn\n";
			for (int i = 0; i < n; ++i) {
				cout << i << '\t' << a[i] << endl;
			}
			c = a[r] + 1;
			a[r] = -1;
			continue;
		}
		++r;
	}
	return 0;
}

File: EightQueen.java
import static java.lang.Math.*;

public class EightQueen
{
	// Eight Queen so n is set to 8
	static int n = 8;
	static int[] a = new int[8];
	static int answers = 0;
	
	public static boolean ok(int r, int c) {
		for (int i = 0; i < r; ++i) {
			if (a[i] == c || (abs(c - a[i]) == abs(r - i))) return false;
		}
		return true;
	}
	
	public static void search(int r) {
		if (r == n) {
			++answers;
			System.out.printf("\nThe NO. %d answer\n", answers);
			System.out.printf("row\tcolumn\n");
			for (int i = 0; i < n; ++i) {
				System.out.printf("%d:\t%d\n", i, a[i]);
			}
		}
		for (int c = 0; c < n; ++c) {
			if (ok(r, c)) {
				a[r] = c;
				search(r + 1);
				a[r] = -1;
			}
		}
	}

	public static void main(String[] args) {
		for (int i = 0; i < n; ++i) {
			a[i] = -1;
		}
		search(0);
	}
}
